apps/prairielearn/src/sync/sharing.ts

```diff
@@ -27,3 +37,99 @@ export async function getInvalidRenames(
   });
   return invalidRenames;
 }
+
+export function getInvalidPublicSharingRemovals(
+  sharedQuestions: SharedQuestion[],
+  courseData: CourseData,
+): string[] {
+  const invalidUnshares: string[] = [];
+  sharedQuestions.forEach((question) => {
+    if (!question.shared_publicly) {
+      return;
+    }
+
+    // TODO: allow if question is not used in anyone else's assessments
+    const questionData = courseData.questions[question.qid].data;
+    if (!(questionData?.sharedPublicly || questionData?.sharedSourcePublicly)) {
+      invalidUnshares.push(question.qid);
+    }
+  });
+  return invalidUnshares;
+}
+
+export async function getInvalidSharingSetDeletions(
+  courseId: string,
+  courseData: CourseData,
+): Promise<string[]> {
+  const sharingSets = await sqldb.queryRows(
+    sql.select_sharing_sets,
+    { course_id: courseId },
+    z.string(),
+  );
+
+  const invalidSharingSetDeletions: string[] = [];
+  const sharingSetNames = (courseData.course.data?.sharingSets || []).map((ss) => ss.name);
+  sharingSets.forEach((sharingSet) => {
+    if (!sharingSetNames.includes(sharingSet)) {
+      invalidSharingSetDeletions.push(sharingSet);
+    }
+  });
+  return invalidSharingSetDeletions;
+}
+
+export function getInvalidSharingSetAdditions(courseData: CourseData): Record<string, string[]> {
```

Comment:
Nit: it'd be helpful to include documentation about what this function returns. Looks like it's a map from QID to sharing set names? Ditto for other functions here.

Reply:
I think the best documentation is actually just looking at what the error message is that gets printed to the user for each of these. Maybe a better solution is to rearrange the code to the error messages are closer to this code? 

If I put separate doc strings on these functions, I feel like I am just putting the same information in 2 different places that we will have to remember to keep in sync later as they are updated.

Reply:
I have an idea. What if I actually just pass in the `logger` to these functions from the parent function? Then the construction of the log error (which is good documentation of what the error is) will be directly in these functions.

Reply:
I made this change, let me know what you think.

Reply:
Ok, this is very annoying. Now I am getting a circular dependency error between `sync/sharing.ts` and `sync/syncFromDisk.ts` even though its just a type import for the `Logger` type. I'm not sure the best solution here.

Reply:
Resolved in 75b4b42.